File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/IgnoreAllFilter.java
package com.codepoetics.raffia.streaming;

import com.codepoetics.raffia.writers.BasketWriter;

import java.math.BigDecimal;

public final class IgnoreAllFilter<T extends BasketWriter<T>> implements FilteringWriter<T> {

  private T target;
  private final FilteringWriter<T> parent;
  private int depth = 0;

  public IgnoreAllFilter(T target, FilteringWriter<T> parent) {
    this.target = target;
    this.parent = parent;
  }

  @Override
  public FilteringWriter<T> beginObject() {
    depth++;
    return this;
  }

  @Override
  public FilteringWriter<T> beginArray() {
    depth++;
    return this;
  }

  @Override
  public FilteringWriter<T> end() {
    return depth-- == 0 ? parent.advance(target) : this;
  }

  @Override
  public FilteringWriter<T> key(String key) {
    return this;
  }

  @Override
  public FilteringWriter<T> add(String value) {
    return this;
  }

  @Override
  public FilteringWriter<T> add(BigDecimal value) {
    return this;
  }

  @Override
  public FilteringWriter<T> add(boolean value) {
    return this;
  }

  @Override
  public FilteringWriter<T> addNull() {
    return this;
  }

  @Override
  public FilteringWriter<T> advance(T newTarget) {
    target = newTarget;
    return this;
  }

  @Override
  public T complete() {
    throw new IllegalStateException("Cannot complete while writing array or object");
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/IgnoreAllFilter.kt
package com.codepoetics.raffia.streaming

import com.codepoetics.raffia.writers.BasketWriter

import java.math.BigDecimal

class IgnoreAllFilter<T : BasketWriter<T>>(private var target: T, private val parent: FilteringWriter<T>) : FilteringWriter<T> {
    private var depth = 0

    override fun beginObject(): FilteringWriter<T> {
        depth++
        return this
    }

    override fun beginArray(): FilteringWriter<T> {
        depth++
        return this
    }

    override fun end(): FilteringWriter<T> {
        return if (depth-- == 0) parent.advance(target) else this
    }

    override fun key(key: String): FilteringWriter<T> {
        return this
    }

    override fun add(value: String): FilteringWriter<T> {
        return this
    }

    override fun add(value: BigDecimal): FilteringWriter<T> {
        return this
    }

    override fun add(value: Boolean): FilteringWriter<T> {
        return this
    }

    override fun addNull(): FilteringWriter<T> {
        return this
    }

    override fun advance(newTarget: T): FilteringWriter<T> {
        target = newTarget
        return this
    }

    override fun complete(): T {
        throw IllegalStateException("Cannot complete while writing array or object")
    }
}
